python/iceberg/hive/hive_tables.py

```diff
@@ -34,12 +42,59 @@ def __init__(self, conf):
     def new_table_ops(self, conf, database, table):
         return HiveTableOperations(conf, self.get_client(), database, table)
 
-    def drop(self, database, table):
-        raise RuntimeError("Not yet implemented")
+    def drop(self, database, table, purge=False) -> None:
+        ops = self.new_table_ops(self.conf, database, table)
+        metadata = ops.current()
 
-    def get_client(self):
+        with self.get_client() as open_client:
+            _logger.info("Deleting {database}.{table} from Hive Metastore".format(database=database, table=table))
+            open_client.drop_table(database, table, deleteData=False)
+
+        if purge:
+            if metadata is not None:
+                manifest_lists_to_delete = []
+                manifests_to_delete = itertools.chain()
+
+                for s in metadata.snapshots:
+                    manifests_to_delete = itertools.chain(manifests_to_delete, (m for m in s.manifests))
+                    if s.manifest_location is not None:
+                        manifest_lists_to_delete.append(s.manifest_location)
+
+                # Make a copy, as it is drained as we explore the manifest to list files.
+                (manifests, manifests_to_delete) = itertools.tee(manifests_to_delete)
+
+                with Pool(self.conf.get(WORKER_THREAD_POOL_SIZE_PROP,
+                          cpu_count())) as delete_pool:
+                    delete_pool.map(self._delete_file(ops), self._unique(self._get_data_files(manifests)))
+                    delete_pool.map(self._delete_file(ops), self._unique(m.manifest_path for m in manifests_to_delete))
+                    delete_pool.map(self._delete_file(ops), self._unique(manifest_lists_to_delete))
+                    delete_pool.map(self._delete_file(ops), [ops.current_metadata_location])
+
+    def get_client(self) -> HMSClient:
         from urllib.parse import urlparse
         metastore_uri = urlparse(self.conf[HiveTables.THRIFT_URIS])
 
         client = hmsclient.HMSClient(host=metastore_uri.hostname, port=metastore_uri.port)
         return client
+
+    def _get_data_files(self, manifests) -> Iterator[str]:
+        return itertools.chain.from_iterable(self._get_data_files_by_manifest(m) for m in manifests)
+
+    def _get_data_files_by_manifest(self, manifest) -> Iterator[str]:
+        file = FileSystemInputFile.from_location(manifest.manifest_path, self.conf)
+        reader = ManifestReader.read(file)
+        return (i.path() for i in reader.iterator())
+
+    @staticmethod
+    def _delete_file(ops):
+        return lambda path: (
+            _logger.info("Deleting file: {path}".format(path=path)),
+            ops.delete_file(path))
+
+    @staticmethod
+    def _unique(iterable: Iterator) -> Iterator:
```

Comment:
I think it might be easier to just catch an exceptoion in `_delete_file`

Reply:
Done